### PR TITLE
[WIP] Tensor decoder for image segmentation

### PR DESCRIPTION
**Pull request by L2zz**
Tensor decoder for image segmentation.

Based on tensordec-directvideo.c and tensordec-boundingbox.c.
Implement skeleton code of tensordec-imagesegment.c
Add meson build for it.

Signed-off-by: niklasjang niklasjang@gmail.com
Signed-off-by: Jihoon Lee ulla4571@gmail.com

### CODE REVIEW

File: ext/nnstreamer/tensor_decoder/tensordec-imagesegment.c [hunk 1]
```diff
@@ -0,0 +1,350 @@
...
+/**
+ * @brief Data structure for image segmentation info.
+ */
+typedef struct
+{
+  image_segment_modes mode; /**< The image segmentation decoding mode */
+  guint **segment_map;  /**< The image segmentated map */
+
+  /*  From option2 */
+  guint o_width; /**< Output video Width */
```
Comment (wooksong) [inline ext/nnstreamer/tensor_decoder/tensordec-imagesegment.c]: NOTE THAT THIS OPTION IS ONLY FOR TFLITE_DEEPLAB_IMAGE_SEGMENT!

File: ext/nnstreamer/tensor_decoder/tensordec-imagesegment.c [hunk 2]
```diff
@@ -0,0 +1,350 @@
...
+#define TFLITE_DEEPLAB_TOTAL_LABELS    21
+#define TFLITE_DEEPLAB_IMAGE_WIDTH     257
+#define TFLITE_DEEPLAB_IMAGE_HEIGHT    257
+
+/**
+ * @brief There can be different schemes for image segmentation.
+ */
+typedef enum
+{
+  TFLITE_DEEPLAB_IMAGE_SEGMENT = 0,
```
Comment (wooksong) [inline ext/nnstreamer/tensor_decoder/tensordec-imagesegment.c]: SEGMENT? SEGMENTATION? 

Actually, this decoder is for image segment(ation), then why do we need the word, image_segment here and there?

File: ext/nnstreamer/tensor_decoder/tensordec-imagesegment.c [hunk 3]
```diff
@@ -0,0 +1,350 @@
...
+is_init (void **pdata)
+{
+  image_segments *idata;
+  *pdata = g_new0 (image_segments, 1);
+
+  idata = *pdata;
+  idata->mode = UNKNOWN_IMAGE_SEGMENT;
+  idata->i_width = 0;
+  idata->i_height = 0;
+  idata->o_width = 0;
```
Comment (wooksong) [inline ext/nnstreamer/tensor_decoder/tensordec-imagesegment.c]: Do we really need o_width and o_height?

File: ext/nnstreamer/tensor_decoder/tensordec-imagesegment.c [hunk 4]
```diff
@@ -0,0 +1,350 @@
...
+    idata->o_width = 0;
+    idata->o_height = 0;
+    if (param == NULL || *param == '\0')
+      return TRUE;
+
+    if (rank < 2) {
+      GST_ERROR
+          ("mode-option-2 of image segmentation is video output dimension (WIDTH:HEIGHT). "
+          "The given parameter, \"%s\", is not acceptable.", param);
+      return TRUE;              /* Ignore this param */
```
Comment (wooksong) [inline ext/nnstreamer/tensor_decoder/tensordec-imagesegment.c]: TRUE?

File: ext/nnstreamer/tensor_decoder/tensordec-imagesegment.c [hunk 5]
```diff
@@ -0,0 +1,350 @@
...
+    idata->i_width = 0;
+    idata->i_height = 0;
+    if (param == NULL || *param == '\0')
+      return TRUE;
+
+    if (rank < 2) {
+      GST_ERROR
+          ("mode-option-3 of image segmentation is input video dimension (WIDTH:HEIGHT). "
+          "The given parameter, \"%s\", is not acceptable.", param);
+      return TRUE;              /* Ignore this param */
```
Comment (wooksong) [inline ext/nnstreamer/tensor_decoder/tensordec-imagesegment.c]: ditto

File: ext/nnstreamer/tensor_decoder/tensordec-imagesegment.c [hunk 6]
```diff
@@ -0,0 +1,350 @@
...
+
+}
+
+/** @brief tensordec-plugin's GstTensorDecoderDef callback */
+static GstFlowReturn
+is_decode (void **pdata, const GstTensorsConfig * config,
+    const GstTensorMemory * input, GstBuffer * outbuf)
+{
+  image_segments *idata = *pdata;
+  const size_t size = idata->i_width * idata->i_height * 4;
```
Comment (wooksong) [inline ext/nnstreamer/tensor_decoder/tensordec-imagesegment.c]: eliminate hard-coded 4. use pre-defined one.

File: ext/nnstreamer/tensor_decoder/tensordec-imagesegment.c (hunk 3 above)
Comment (L2zz) [reply ext/nnstreamer/tensor_decoder/tensordec-imagesegment.c]: May be not. I'll remove it

File: ext/nnstreamer/tensor_decoder/tensordec-imagesegment.c (hunk 4 above)
Comment (L2zz) [reply ext/nnstreamer/tensor_decoder/tensordec-imagesegment.c]: TRUE means just ignore wrong option and keep going pipeline.

File: ext/nnstreamer/tensor_decoder/tensordec-imagesegment.c (hunk 6 above)
Comment (L2zz) [reply ext/nnstreamer/tensor_decoder/tensordec-imagesegment.c]: Yes, I'll change it

File: ext/nnstreamer/tensor_decoder/tensordec-imagesegment.c (hunk 2 above)
Comment (L2zz) [reply ext/nnstreamer/tensor_decoder/tensordec-imagesegment.c]: I feel it is too long to use it. If it is hard to understand, I'll change it.

File: ext/nnstreamer/tensor_decoder/tensordec-imagesegment.c (hunk 1 above)
Comment (L2zz) [reply ext/nnstreamer/tensor_decoder/tensordec-imagesegment.c]: Sry, I don't know what to do for it.

Comment (L2zz) [reply ext/nnstreamer/tensor_decoder/tensordec-imagesegment.c]: Does it mean do not place the option dependent value in common data structure?

File: ext/nnstreamer/tensor_decoder/tensordec-imagesegment.c (hunk 3 above)
Comment (L2zz) [reply ext/nnstreamer/tensor_decoder/tensordec-imagesegment.c]: Do we need to get i_width and i_height by option?

**Review by wooksong — CHANGES_REQUESTED** [review]
Need to improve

File: ext/nnstreamer/tensor_decoder/tensordec-imagesegment.c [hunk 7]
```diff
@@ -0,0 +1,287 @@
...
+}
+
+/** @brief tensordec-plugin's GstTensorDecoderDef callback */
+static int
+is_setOption (void **pdata, int op_num, const char *param)
+{
+  image_segments *idata = *pdata;
+
+  if (op_num == 0) {
+    /* option1 = image segmentation decoder mode */
```
Comment (wooksong) [inline ext/nnstreamer/tensor_decoder/tensordec-imagesegment.c]: option == 0 and option1 keep confusing me.
```
The first option indicates a mode of this decoder sub-plugin.
```

Or you use your own phrase. Anyway change it.

File: ext/nnstreamer/tensor_decoder/tensordec-imagesegment.c [hunk 8]
```diff
@@ -0,0 +1,287 @@
...
+#include <string.h>
+#include <glib.h>
+#include <gst/video/video-format.h>
+#include <nnstreamer_plugin_api_decoder.h>
+#include <nnstreamer_plugin_api.h>
+
+void init_is (void) __attribute__ ((constructor));
+void fini_is (void) __attribute__ ((destructor));
+
+#define RGBA_CHANNEL                   4
```
Comment (wooksong) [inline ext/nnstreamer/tensor_decoder/tensordec-imagesegment.c:70]: Is this only work for RGBA?

File: ext/nnstreamer/tensor_decoder/tensordec-imagesegment.c [hunk 9]
```diff
@@ -0,0 +1,287 @@
...
+#define TFLITE_DEEPLAB_TOTAL_LABELS    21
```
Comment (wooksong) [inline ext/nnstreamer/tensor_decoder/tensordec-imagesegment.c:71]: ditto

File: ext/nnstreamer/tensor_decoder/tensordec-imagesegment.c [hunk 10]
```diff
@@ -0,0 +1,287 @@
...
+
+/**
+ * @brief There can be different schemes for image segmentation
+ */
+typedef enum
+{
+  TFLITE_DEEPLAB = 0,
+  UNKNOWN,
```
Comment (wooksong) [inline ext/nnstreamer/tensor_decoder/tensordec-imagesegment.c]: MODE_TFILITE_DEEPLAB
MODE_UNKNOWN

File: ext/nnstreamer/tensor_decoder/tensordec-imagesegment.c [hunk 11]
```diff
@@ -0,0 +1,287 @@
...
+      return FALSE;
+    }
+
+    if (idata->mode != previous && idata->mode != UNKNOWN) {
+      return TRUE;
+    }
+    return TRUE;
+  }
+
+  GST_WARNING ("mode-option-\"%d\" is not definded.", op_num);
```
Comment (wooksong) [inline ext/nnstreamer/tensor_decoder/tensordec-imagesegment.c:168]: mode-option or just mode?

File: ext/nnstreamer/tensor_decoder/tensordec-imagesegment.c [hunk 12]
```diff
@@ -0,0 +1,287 @@
...
+  }
+
+  set_color_according_to_label (idata, &out_info);
+
+  gst_memory_unmap (out_mem, &out_info);
+
+  if (gst_buffer_get_size (outbuf) == 0)
+    gst_buffer_append_memory (outbuf, out_mem);
+
+  return GST_FLOW_OK;
```
Comment (wooksong) [inline ext/nnstreamer/tensor_decoder/tensordec-imagesegment.c:288]: Q: Does this function always return GST_FLOW_OK unless it dies with assertion?

File: ext/nnstreamer/tensor_decoder/tensordec-imagesegment.c (hunk 11 above)
Comment (L2zz) [reply ext/nnstreamer/tensor_decoder/tensordec-imagesegment.c:168]: Mode-option, if user write invalid option like option3(we don't have), then warn that case.

File: ext/nnstreamer/tensor_decoder/tensordec-imagesegment.c (hunk 12 above)
Comment (L2zz) [reply ext/nnstreamer/tensor_decoder/tensordec-imagesegment.c:288]: Yes

**Pull request closed without merging**